airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/DefaultTyperDeduper.java

```diff
@@ -55,43 +56,29 @@ public void prepareFinalTables() throws Exception {
     if (overwriteStreamsWithTmpTable != null) {
       throw new IllegalStateException("Tables were already prepared.");
     }
-    overwriteStreamsWithTmpTable = new HashMap<>();
+    overwriteStreamsWithTmpTable = new HashSet<>();
 
     // For each stream, make sure that its corresponding final table exists.
     // Also, for OVERWRITE streams, decide if we're writing directly to the final table, or into an
     // _airbyte_tmp table.
     for (StreamConfig stream : parsedCatalog.streams()) {
       final Optional<DialectTableDefinition> existingTable = destinationHandler.findExistingTable(stream.id());
-      if (existingTable.isEmpty()) {
-        // The table doesn't exist. Create it.
-        destinationHandler.execute(sqlGenerator.createTable(stream, NO_SUFFIX));
-        if (stream.destinationSyncMode() == DestinationSyncMode.OVERWRITE) {
-          // We're creating this table for the first time. Write directly into it.
-          overwriteStreamsWithTmpTable.put(stream.id(), NO_SUFFIX);
+      if (existingTable.isPresent()) {
+        // The table already exists. Decide whether we're writing to it directly, or using a tmp table.
+        if (stream.destinationSyncMode() == DestinationSyncMode.OVERWRITE && !destinationHandler.isFinalTableEmpty(stream.id())) {
+          // We want to overwrite an existing table. Write into a tmp table. We'll overwrite the table at the
+          // end of the sync.
+          overwriteStreamsWithTmpTable.add(stream.id());
+          destinationHandler.execute(sqlGenerator.createTable(stream, TMP_OVERWRITE_TABLE_SUFFIX));
         }
-      } else {
-        // The table exists. Make sure it has the right schema.
-        if (!sqlGenerator.existingSchemaMatchesStreamConfig(stream, existingTable.get())) {
+        if (!overwriteStreamsWithTmpTable.contains(stream.id()) && !sqlGenerator.existingSchemaMatchesStreamConfig(stream, existingTable.get())) {
```

Comment:
I think you could technically do `else if (!sqlGenerator.existingSchemaMatchesStreamConfig(stream, existingTable.get()))` here? I think that still correctly covers case 1 and 2 from the other comment thread.

Reply:
oh yeah that's a good point

Reply:
unit test says this works. turning on automerge